src/H5Dchunk.c

```diff
@@ -5763,48 +5609,24 @@ H5D__chunk_collective_fill(const H5D_t *dset, H5D_chunk_coll_fill_info_t *chunk_
     if (H5CX_set_io_xfer_mode(H5FD_MPIO_COLLECTIVE) < 0)
         HGOTO_ERROR(H5E_DATASET, H5E_CANTSET, FAIL, "can't set transfer mode");
 
-    /* Low-level write (collective) */
-    if (H5F_shared_block_write(H5F_SHARED(dset->oloc.file), H5FD_MEM_DRAW, (haddr_t)0,
-                               (blocks) ? (size_t)1 : (size_t)0, fill_buf) < 0)
-        HGOTO_ERROR(H5E_IO, H5E_WRITEERROR, FAIL, "unable to write raw data to file");
-
-    /* Barrier so processes don't race ahead */
-    if (MPI_SUCCESS != (mpi_code = MPI_Barrier(mpi_comm)))
-        HMPI_GOTO_ERROR(FAIL, "MPI_Barrier failed", mpi_code)
+    if (H5F_shared_vector_write(H5F_SHARED(dset->oloc.file), io_count, io_types, io_addrs,
+                                all_same_block_len ? io_2sizes : io_sizes, io_wbufs) < 0)
+        HGOTO_ERROR(H5E_DATASET, H5E_WRITEERROR, FAIL, "vector write call failed");
 
 done:
     if (have_xfer_mode)
         /* Set transfer mode */
         if (H5CX_set_io_xfer_mode(prev_xfer_mode) < 0)
             HDONE_ERROR(H5E_DATASET, H5E_CANTSET, FAIL, "can't set transfer mode");
 
-    /* free things */
-    if (MPI_BYTE != file_type)
-        if (MPI_SUCCESS != (mpi_code = MPI_Type_free(&file_type)))
-            HMPI_DONE_ERROR(FAIL, "MPI_Type_free failed", mpi_code)
-    if (MPI_BYTE != mem_type)
-        if (MPI_SUCCESS != (mpi_code = MPI_Type_free(&mem_type)))
-            HMPI_DONE_ERROR(FAIL, "MPI_Type_free failed", mpi_code)
-    H5MM_xfree(chunk_disp_array);
-    H5MM_xfree(block_disps);
-    H5MM_xfree(block_lens);
+    H5MM_xfree(io_addrs);
+    H5MM_xfree(io_wbufs);
+    if (!all_same_block_len)
+        H5MM_xfree(io_sizes);
```

Comment:
Minor comment: I'd prefer to just always call H5MM_xfree here to make the code more robust against future changes. It's a no-op if io_sizes in NULL so this won't cause problems

Reply:
The C standard explicitly says that calling free on NULL pointers is acceptable.

Reply:
Fixed